components/eam/src/physics/cam/hetfrz_classnuc_cam.F90

```diff
@@ -876,368 +890,383 @@ end subroutine hetfrz_classnuc_cam_save_cbaero
-      else
-         awfacm(2) = 0._r8
-      end if
 
-      ! accumulation mode for bc (if MAM4, primary carbon mode is insoluble)
-      if (aer(ii,kk,num_accum) > 0._r8) then
-         awcam(1) = (bc_num*1.0e6_r8)/aer(ii,kk,num_accum)* &
-            ( aer(ii,kk,so4_accum) + aer(ii,kk,soa_accum) + aer(ii,kk,pom_accum) + aer(ii,kk,bc_accum) + &
-              aer(ii,kk,mom_accum) )*1.0e9_r8 ! [mug m-3]
-      else
-         awcam(1) = 0._r8
-      end if
-      awfacm(1) = awfacm(2)
+   vol_core(2) = aer(ii,kk,dst_accum)/(specdens_dust*rhoair)
 
-      ! coarse mode for dust_a3
-      if (aer(ii,kk,num_coarse) > 0._r8) then
-         awcam(3) = (dst3_num*1.0e6_r8)/aer(ii,kk,num_coarse)* ( aer(ii,kk,so4_coarse) + & 
-                     aer(ii,kk,mom_coarse) + aer(ii,kk,bc_coarse) + aer(ii,kk,pom_coarse) + aer(ii,kk,soa_coarse) ) *1.0e9_r8
-      else
-         awcam(3) = 0._r8
-      end if
+   ! ratio1 = vol_shell/vol_core = actual hygroscopic-shell-volume/dust-core-volume
+   ! ratio2 = 6.0_r8*dr_so4_monolayers_pcage/(dgncur_a*fac_volsfc_dust)
+   !        = (shell-volume corresponding to n_so4_monolayers_pcage)/core-volume 
+   !   
+   ! The 6.0/(dgncur_a*fac_volsfc_dust) = (mode-surface-area/mode-volume)
+   ! Note that vol_shell includes both so4, pom, AND soa as "equivalent so4",
+   ! The soa_equivso4_factor accounts for the lower hygroscopicity of soa.
+   !
+   ! Define xferfrac_pcage = min(1.0, ratio1/ratio2)
+   !   But ratio1/ratio2 == tmp1/tmp2, and coding below avoids possible overflow 
+
+   ! bc
+   fac_volsfc_bc      = exp(2.5_r8*alnsg_mode_pcarbon**2)
+
+   vol_shell(1) = ( aer(ii,kk,pom_pcarbon)*pom_equivso4_factor/specdens_pom + &
+                    aer(ii,kk,mom_pcarbon)*mom_equivso4_factor/specdens_mom )/rhoair
+
+   vol_core(1)  = aer(ii,kk,bc_pcarbon)/(specdens_bc*rhoair)
+   coat_ratio1 = vol_shell(1)*(r_bc*2._r8)*fac_volsfc_bc
+   coat_ratio2 = max(6.0_r8*dr_so4_monolayers_dust*vol_core(1), 0.0_r8)
+   dstcoat(1) = coat_ratio1/coat_ratio2
+
+   ! dust_a1
+   coat_ratio1 = vol_shell(2)*(r_dust_a1*2._r8)*fac_volsfc_dust_a1
+   coat_ratio2 = max(6.0_r8*dr_so4_monolayers_dust*vol_core(2), 0.0_r8)
+   dstcoat(2) = coat_ratio1/coat_ratio2
+
+   ! dust_a3
+   vol_shell(3) = aer(ii,kk,so4_coarse)/(specdens_so4*rhoair) + & 
+                  aer(ii,kk,pom_coarse)/(specdens_pom*rhoair) + & 
+                  aer(ii,kk,soa_coarse)/(specdens_soa*rhoair) + & 
+                  aer(ii,kk,mom_coarse)/(specdens_mom*rhoair)
+
+   vol_core(3)  = aer(ii,kk,dst_coarse)/(specdens_dust*rhoair)
+   coat_ratio1 = vol_shell(3)*(r_dust_a3*2._r8)*fac_volsfc_dust_a3
+   coat_ratio2 = max(6.0_r8*dr_so4_monolayers_dust*vol_core(3), 0.0_r8)
+   dstcoat(3) = coat_ratio1/coat_ratio2
+      
+   dstcoat(1) = min_max_bound(0.001_r8, 1.0_r8, dstcoat(1))
+   dstcoat(2) = min_max_bound(0.001_r8, 1.0_r8, dstcoat(2))
+   dstcoat(3) = min_max_bound(0.001_r8, 1.0_r8, dstcoat(3))
+
+
+   do ispec = 1, 3
```

Comment:
Is it per species or should it be `imode` instead of `ispec`?

Reply:
It's for bc in accum mode, dust in accum mode, dust in coarse mode. ispec should be good enough.